sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala

```diff
@@ -81,6 +81,14 @@ object LogicalRelation {
     // The v1 source may return schema containing char/varchar type. We replace char/varchar
     // with "annotated" string type here as the query engine doesn't support char/varchar yet.
     val schema = CharVarcharUtils.replaceCharVarcharWithStringInSchema(relation.schema)
-    LogicalRelation(relation, schema.toAttributes, Some(table), false)
+    val attributes = schema.toAttributes
+    val output = table.stats.map(_.colStats) match {
+      case Some(colStats) =>
+        attributes
+          .map(a => a.withNullability(colStats.get(a.name).forall(_.nullCount.forall(_ > 0L))))
```

Comment:
Hmmm .. does this mean that if a table is saved with nullable schema but the table doesn't have nulls, here the schema becomes non-nullable when we read it back?

Reply:
Yes. Based on column statistics.

Reply:
> Based on column statistics.

Stats doesn't have to be 100% accurate?
- it could be an estimate based on sampled data;
- it could be outdated with the ground truth data.

Reply:
I have the same feeling with @HyukjinKwon @cloud-fan ...